src/network/networkd-varlink.c

```diff
@@ -1,6 +1,158 @@
+
+        l = fread(&u, 1, sizeof(u), f);
+        if (l == 0 && feof(f))
+                return 0;
+        if (l != sizeof(u))
+                return -EBADMSG;
+
+        /* each LLDP packet is at most MTU size, but let's allow up to 4KiB just in case */
+        if (le64toh(u) >= 4096)
+                return -EBADMSG;
+
+        raw = new(uint8_t, le64toh(u));
+        if (!raw)
+                return -ENOMEM;
+
+        if (fread(raw, 1, le64toh(u), f) != le64toh(u))
+                return -EBADMSG;
+
+        r = sd_lldp_neighbor_from_raw(ret, raw, le64toh(u));
+        if (r < 0)
+                return r;
+
+        return 1;
+}
+
+static int dump_lldp_neighbors(int ifindex, Varlink *link) {
+        _cleanup_fclose_ FILE *f = NULL;
+        _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+        int r;
+
+        assert(ifindex > 0);
+        assert(link);
+
+        r = open_lldp_neighbors(ifindex, &f);
+        if (r == -ENOENT) {
+                return varlink_replyb(link, JSON_BUILD_EMPTY_ARRAY);
+        }
+        if (r < 0)
+                return r;
+
+        for (;;) {
+                const char *chassis_id = NULL, *port_id = NULL, *system_name = NULL, *port_description = NULL;
+                _cleanup_(sd_lldp_neighbor_unrefp) sd_lldp_neighbor *n = NULL;
+                uint16_t cc;
+
+                r = next_lldp_neighbor(f, &n);
+                if (r < 0) {
+                        log_warning_errno(r, "Failed to read neighbor data: %m");
+                        break;
+                }
+                if (r == 0)
+                        break;
+
+                if (v) {
+                        r = varlink_notify(link, v);
+                        if (r < 0)
+                                return r;
+
+                        v = json_variant_unref(v);
+                }
+
+                (void) sd_lldp_neighbor_get_chassis_id_as_string(n, &chassis_id);
+                (void) sd_lldp_neighbor_get_port_id_as_string(n, &port_id);
+                (void) sd_lldp_neighbor_get_system_name(n, &system_name);
+                (void) sd_lldp_neighbor_get_port_description(n, &port_description);
+
+                r = sd_lldp_neighbor_get_enabled_capabilities(n, &cc);
+
+                r = json_build(&v, JSON_BUILD_OBJECT(
+                                JSON_BUILD_PAIR("ChassisId", JSON_BUILD_STRING(chassis_id)),
+                                JSON_BUILD_PAIR("SystemName", JSON_BUILD_STRING(system_name)),
+                                JSON_BUILD_PAIR_CONDITION(r >= 0, "EnabledCapabilities", JSON_BUILD_UNSIGNED(cc)),
+                                JSON_BUILD_PAIR("PortId", JSON_BUILD_STRING(port_id)),
+                                JSON_BUILD_PAIR_CONDITION(port_description, "PortDescription", JSON_BUILD_STRING(port_description))));
+
+                if (r < 0)
+                        return r;
+        }
+
+        if (!v)
+                return varlink_replyb(link, JSON_BUILD_EMPTY_ARRAY);
```

Comment:
so you return an object if there's an entry and an empty array if not. that's weird. please either return a valid object as suggested above, or some error code, i.e. io.systemd.Network.NoNeighborFound or so

Reply:
I'm surprised by this suggestion, and I don't think that "no LLDP peers on this link" is an error. These data come from a network, and whether there is an LLDP-speaking neighbor or not, that is going to depend on my particular network.

In my opinion, the response should be a JSON array of objects, one object per each neighbor on a given link. If you special-case an empty array to an error code, that will only make the code more complex (on both sides of the JSON generation and parsing). What advantages do you see in encoding "no neighbors here" as a special value?

Reply:
generating a json list object of the reply is problematic, since it means we cannot stream the response that naturally. typically json arrays are kept as one thing in memory and passed around as one ting. This is fine if the size is reasonably bounded. But a list of peers is not. So there we use the varlink concept that allows replying to method calls with continuations: i.e. send an initial reply with the first object, then a second continuation reply, with another object and so on. The last object is then marked as final. This concept is built into the varlink concept, and makes the protocol "streamable": you can theoretically stream a completely unbounded number of objects through a varlink IPC connection, very much unlike on dbus or other such IPCs.

So that's the reason why instead of passing a list we pass multiple replies with one object each. That leaves one question open: what to do when there's no object to return. We could return an empty object, but that sucks I think. I think its a bit better to return "Not objects matching your request" as an error indication, i.e. out of band.

There's prior art for that in our codebase: userdbd returns an error if you look for a specific user/group record or a number of them and there are no matching entries.

Reply:
Ah, thanks, I was focusing on the "final JSON shape" and was not considering the streaming use case. Now, if the streaming thing needs to push chunks of JSON which are still valid JSON files on their own, then I agree with your comment.

Just FYI, in the YANG modeling language there's the [`empty` type](https://datatracker.ietf.org/doc/html/rfc7950#section-9.11) which is [modeled as `[null]`](https://datatracker.ietf.org/doc/html/rfc7951#section-6.9). I'm not terribly happy with that one, but I'm adding that as a data point. I agree that some explicit value would work just as well (and it's JSON, so perhaps a magic string is the way to go?).

Reply:
Yes, this was an obvious error on my end. Thanks for catching this, I will reply with `varlink_error`.